Sources/Basics/ProgressAnimation/NinjaProgressAnimation.swift

```diff
@@ -0,0 +1,101 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import class TSCBasic.TerminalController
+import protocol TSCBasic.WritableByteStream
+
+extension ProgressAnimation {
+    /// A ninja-like progress animation that adapts to the provided output stream.
+    @_spi(SwiftPMInternal_ProgressAnimation)
+    public static func ninja(
+        stream: WritableByteStream,
+        verbose: Bool
+    ) -> any ProgressAnimationProtocol {
+        Self.dynamic(
+            stream: stream,
+            verbose: verbose,
+            ttyTerminalAnimationFactory: { RedrawingNinjaProgressAnimation(terminal: $0) },
+            dumbTerminalAnimationFactory: { SingleLinePercentProgressAnimation(stream: stream, header: nil) },
+            defaultAnimationFactory: { MultiLineNinjaProgressAnimation(stream: stream) })
+    }
+}
+
+/// A redrawing ninja-like progress animation.
+final class RedrawingNinjaProgressAnimation: ProgressAnimationProtocol {
```

Comment:
IMO the original types in TSC should be deprecated so that contributors without enough context don't assume that the TSC version will be maintained going forward.

Reply:
I can do that after we land this PR, does that sound good?